libvast/vast/system/make_source.hpp

```diff
@@ -0,0 +1,181 @@
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/command.hpp"
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/endpoint.hpp"
+#include "vast/concept/parseable/vast/expression.hpp"
+#include "vast/concept/parseable/vast/schema.hpp"
+#include "vast/defaults.hpp"
+#include "vast/detail/make_io_stream.hpp"
+#include "vast/detail/string.hpp"
+#include "vast/endpoint.hpp"
+#include "vast/error.hpp"
+#include "vast/expression.hpp"
+#include "vast/fwd.hpp"
+#include "vast/logger.hpp"
+#include "vast/schema.hpp"
+#include "vast/system/accountant.hpp"
+#include "vast/system/datagram_source.hpp"
+#include "vast/system/signal_monitor.hpp"
+#include "vast/system/source.hpp"
+#include "vast/system/type_registry.hpp"
+
+#include <caf/actor.hpp>
+#include <caf/actor_cast.hpp>
+#include <caf/actor_system.hpp>
+#include <caf/io/middleman.hpp>
+#include <caf/settings.hpp>
+#include <caf/spawn_options.hpp>
+
+namespace vast::system {
+
+namespace {
+
+caf::expected<expression> parse_expression(command::argument_iterator begin,
+                                           command::argument_iterator end) {
+  auto str = detail::join(begin, end, " ");
+  auto expr = to<expression>(str);
+  if (expr)
+    expr = normalize_and_validate(*expr);
+  return expr;
+}
+
+} // namespace
+
+/// Tries to spawn a new SOURCE for the specified format.
+/// @tparam Reader the format-specific reader.
+/// @tparam Defaults defaults for the format-specific reader.
+/// @tparam SpawnOptions caf::spawn_options to pass to sys.spawn().
+/// @param self Points to the parent actor.
+/// @param sys The actor system to spawn the source in.
+/// @param inv The invocation that prompted the actor to be spawned.
+/// @param accountant A handle to the accountant component.
+/// @param type_registry A handle to the type registry component.
+/// @param importer A handle to the importer component.
+/// @returns a handle to the spawned actor on success, an error otherwise
+template <class Reader, class Defaults,
+          caf::spawn_options SpawnOptions = caf::spawn_options::no_flags,
+          class Actor>
+caf::expected<caf::actor>
+make_source(const Actor& self, caf::actor_system& sys, const invocation& inv,
```

Comment:
`self` is only used twice at the end of the function, we could either move these uses into the calling context or use `caf::anon_send`. That would at least allow us to get rid of the `Actor` template parameter.

Reply:
I'd rather not do this. `anon_send` needlessly obfuscates the logs, and I also don't want to spread out the initialization of the source again. After all, the whole point of this function is to unify the configuration.

Reply:
I'll cede my case in this particular instance. I want to repeat tough, that we won't measure any runtime advantage from monomorphing the code paths around spawning sources and sinks, but it does cost us quite a bit in development time.